packages/core/src/combinator/timeslice.js

```diff
@@ -3,111 +3,102 @@
 /** @author John Hann */
 
 import Pipe from '../sink/Pipe'
-import { disposeAll } from '@most/disposable'
+import { disposeBoth } from '@most/disposable'
 import { join } from './chain'
+import { merge } from './merge'
+import { never } from '../source/never'
+
+const timeBounds = (min, minSignal, maxSignal) =>
+  ({ min, minSignal, maxSignal })
+
+const minBounds = (minSignal) =>
+  timeBounds(Infinity, minSignal, never())
+
+const maxBounds = (maxSignal) =>
+  timeBounds(0, never(), maxSignal)
+
+const mergeTimeBounds = (b1, b2) =>
```

Comment:
Time bounds are a representation of the start/end bounds of a time slice:

* `min` is a known minimum time, before which events will be skipped.
* `minSignal` is a signal which can provide an update the minimum time
* `maxSignal` is a signal that represents the future maximum time after which events will be discarded, and the time sliced stream will end.

Merging time bounds takes the _later_ min value (i.e. min can only be "narrowed", and `Infinity` annihilates min values), and _merges_ signals.  Merging signals means the earliest event will win, which is what we want.

Reply:
> Merging signals means the earliest event will win, which is what I think we want.

Aha, wrong again!

That's correct for the maxSignal, but not for minSignal.  I think we want both only to narrow.  That is, I think we want the "earlier first event" of two maxSignals but the "later first event" of the two minSignals.

That's a bit trickier to implement.  Probably time for a little TDD to ensure I get that right.

Reply:
Ok, 1809ca3 has an algorithm for narrowing minSignals (i.e. "latest first event").  The algorithm is lazy in that merging time bounds simply concatenates arrays of signal streams (a la free monoid).  Then, later in `Timeslice#run`, `getMinSignal` does the work of folding that array of min signals into a single stream containing the "latest first event" using `take` and `slice`.